File: conformance/livewire/app/Store.php
<?php

namespace App;

/**
 * In-memory todo store for the Faultsense Livewire conformance harness.
 *
 * This harness runs Laravel's built-in development server (`php artisan
 * serve`), which serves requests through PHP-FPM-style short-lived
 * processes. That means static class state does NOT survive across
 * requests. We therefore persist the store to a tiny JSON file on
 * every mutation — the file lives under /tmp inside the container and
 * is wiped when the Playwright driver hits POST /todos/reset at the
 * start of each test.
 *
 * Kept as a plain static class (no Eloquent, no database) so the
 * harness has zero schema setup and maps one-to-one to the
 * conformance/hotwire/ Store class.
 */
class Store
{
    private const STATE_FILE = '/tmp/livewire-harness-state.json';

    public static function all(): array
    {
        return self::load()['todos'];
    }

    public static function active(): bool
    {
        return self::load()['active'];
    }

    public static function setActive(bool $value): void
    {
        $state = self::load();
        $state['active'] = $value;
        self::save($state);
    }

    public static function add(string $text): array
    {
        $state = self::load();
        $todo = [
            'id' => $state['next_id'],
            'text' => $text,
            'completed' => false,
        ];
        $state['next_id']++;
        $state['todos'][] = $todo;
        self::save($state);
        return $todo;
    }

    public static function toggle(int $id): void
    {
        $state = self::load();
        foreach ($state['todos'] as &$todo) {
            if ($todo['id'] === $id) {
                $todo['completed'] = !$todo['completed'];
                break;
            }
        }
        unset($todo);
        self::save($state);
    }

    public static function remove(int $id): void
    {
        $state = self::load();
        $state['todos'] = array_values(array_filter(
            $state['todos'],
            fn (array $t): bool => $t['id'] !== $id
        ));
        self::save($state);
    }

    public static function reset(): void
    {
        self::save(['todos' => [], 'next_id' => 1, 'active' => false]);
    }

    private static function load(): array
    {
        if (!file_exists(self::STATE_FILE)) {
            return ['todos' => [], 'next_id' => 1, 'active' => false];
        }
        $raw = file_get_contents(self::STATE_FILE);
        $decoded = json_decode($raw, true);
        return is_array($decoded)
            ? $decoded
            : ['todos' => [], 'next_id' => 1, 'active' => false];
    }

    private static function save(array $state): void
    {
        file_put_contents(self::STATE_FILE, json_encode($state));
    }
}

File: conformance/liveview/lib/faultsense/store.ex
defmodule Faultsense.Store do
  @moduledoc """
  In-memory todo store. Backed by an `Agent` so all LiveView processes
  share the same state and the Playwright driver can wipe it via the
  POST /reset endpoint between scenarios.

  Mirrors conformance/hotwire/app/controllers/todos_controller.rb's
  Store class one-to-one so the shared driver runners see the same
  affordances (ids, completed flag, active flag).
  """

  use Agent

  @type todo :: %{id: integer(), text: String.t(), completed: boolean()}

  @initial_state %{todos: [], next_id: 1, active: false}

  def start_link(_opts) do
    Agent.start_link(fn -> @initial_state end, name: __MODULE__)
  end

  @spec all() :: [todo()]
  def all, do: Agent.get(__MODULE__, & &1.todos)

  @spec active?() :: boolean()
  def active?, do: Agent.get(__MODULE__, & &1.active)

  @spec set_active(boolean()) :: :ok
  def set_active(value) do
    Agent.update(__MODULE__, &Map.put(&1, :active, value))
  end

  @spec add(String.t()) :: todo()
  def add(text) do
    Agent.get_and_update(__MODULE__, fn state ->
      todo = %{id: state.next_id, text: text, completed: false}

      new_state =
        state
        |> Map.update!(:todos, &(&1 ++ [todo]))
        |> Map.update!(:next_id, &(&1 + 1))

      {todo, new_state}
    end)
  end

  @spec toggle(integer()) :: :ok
  def toggle(id) do
    Agent.update(__MODULE__, fn state ->
      todos =
        Enum.map(state.todos, fn
          %{id: ^id} = todo -> %{todo | completed: not todo.completed}
          other -> other
        end)

      %{state | todos: todos}
    end)
  end

  @spec remove(integer()) :: :ok
  def remove(id) do
    Agent.update(__MODULE__, fn state ->
      %{state | todos: Enum.reject(state.todos, &(&1.id == id))}
    end)
  end

  @spec reset() :: :ok
  def reset do
    Agent.update(__MODULE__, fn _state -> @initial_state end)
  end
end
